File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/strongly_connected_components.cpp
#include "graphs.h"

void dfsRec(adjacencyMap(int) g, int startNode, visitedMap(int) &visited, vector<int> &order) {
    visited[startNode] = true;
    for (int neighbour : g[startNode]) {
        if (!visited[neighbour]) {
            dfsRec(g, neighbour, visited, order);
        }
    }
    order.push_back(startNode);
}

// initVisited maps all nodes as unvisited.
visitedMap(int) initVisitedInt(adjacencyMap(int) graph) {
    visitedMap(int) visited;
    for (auto const&[node, neighbous] : graph) {
        visited[node] = false;
    }
    return visited;
}

// Topological sorting using DFS.
vector<int> dfsTopoSort(adjacencyMap(int) g){
    visitedMap(int) visited = initVisitedInt(g);
    vector<int> order = {};

    for (auto const&[node, _] : g) {
        if (!visited[node]) {
            dfsRec(g, node, visited, order);
        }
    }

    reverse(order.begin(), order.end());
    return order;
}

// Given a digraph, returns a digraph with the direction of all edges inverted.
adjacencyMap(int) reverseGraph(adjacencyMap(int) g) {
    adjacencyMap(int) revGraph; for(auto const&[node, _] : g) revGraph[node] = {};
    for(auto const&[node, neighbours] : g) {
        for (int neighbour : neighbours) {
            revGraph[neighbour].push_back(node);
        }
    }

    return revGraph;
}

void compDfs(adjacencyMap(int) g, int startNode, visitedMap(int) &visited, unordered_map<int, int> &components, int &c) {
    visited[startNode] = true;
    components[startNode] = c;

    for(int neighbour : g[startNode]) {
        if (!visited[neighbour]) {
            compDfs(g, neighbour, visited, components, c);
        }
    }
}

// Kosaraju-Shamir algorithm to find strongly connected components in a Digraph.
unordered_map<int, int> kosarajuShamir(adjacencyMap(int) g) {
    adjacencyMap(int) revGraph = reverseGraph(g);
    vector<int> order = dfsTopoSort(revGraph);

    visitedMap(int) visited = initVisitedInt(g);
    unordered_map<int, int> components = {};
    int c = 0;

    for(int node : order) {
        if (!visited[node]) {
            compDfs(g, node, visited, components, c);
            c++;
        }
    }

    return components;
} 

int main(void) {
    for (auto const&[node, comp] : kosarajuShamir(G4)) cout << node << ": " << comp << "\n";
    return EXIT_SUCCESS;
}

File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/strongly_connected_components.py
from graphs import G4, init_visited
from toposrt import dfs_toposort


def reverse_graph(g: dict[int, list[int]]) -> dict[int, list[int]]:
    """Given a digraph, returns a digraph with the direction of all edges inverted."""
    rev_g: dict[int, list[int]] = {node: [] for node in g.keys()}
    for node, neighbours in g.items():
        for neighbour in neighbours:
            rev_g[neighbour].append(node)
    return rev_g


def kosaraju_shamir(g: dict[int, list[int]]) -> dict[int, int]:
    """Kosaraju-Shamir algorithm to find strongly connected components in a Digraph."""
    rev_g: dict[int, list[int]] = reverse_graph(g)
    order: list[int] = dfs_toposort(g=rev_g)

    visited: dict[int, bool] = init_visited(graph=g)
    components: dict[int, int] = {}
    c: int = 0

    def dfs(g: dict[int, list[int]], start_node: int, visited: dict[int, bool], components: dict[int, int], c: int):
        visited[start_node] = True
        components[start_node] = c
        for neighbour in g[start_node]:
            if not visited[neighbour]:
                dfs(g=g, start_node=neighbour, visited=visited, components=components, c=c)

    for node in order:
        if not visited[node]:
            dfs(g=g, start_node=node, visited=visited, components=components, c=c)
            c += 1

    return components


def main():
    print(kosaraju_shamir(G4))


if __name__ == "__main__":
    main()
